server/renderer.js

```diff
@@ -1,363 +1,172 @@
-			if (m) {
-				var context = extend(true, {}, this.routes[i])
-				if (context.cache === false) {
-					return new Promise(function (resolve, reject) {resolve()})	
-				}
-				for (var ind = 1; ind < m.length; ind++) {
-					if (context.groups && (ind-1) < context.groups.length) {
-						context[context.groups[ind-1]] = m[ind]
-					}
+			.catch(() => { // it's not cached, so we'll render it
+				let route = routes.find((route) => route.path.test(path));
+
+				if (route === undefined) {
+					logger.warn(path);
+					this.fourohfour(undefined, res, undefined);
+					return;
 				}
-				var out = this.renderPath(context)
-				var written = RENDER_ROOT_STR + url.replace(/\//g,".")
-				logger.info("[render]", "Caching", url)
-				return denodeify(fs.writeFile, [path.join(loc, written), out])
-			}
-		}
-	},
-	handle: function(req, res, next) {
-		var that = this;
-		for (var i in this.routes) {
-			var m = req.originalUrl.match(i)
-			if (m && req.method == "GET") {
-				var context = extend(true, {}, this.routes[i])
-				if (context.cache === true) {
-					var written = RENDER_ROOT_STR + req.originalUrl.replace(/\//g,".")
-					res.type(context.mime || "text/html")
-					promiseFile(path.join(this.__dirname, config.paths.render), written).then(function(val) {
-						res.send(val.data)
-					}, function(err) {
-						that.fourohfour(req, res, next)
-					}).catch(crash)
-					return 
-				} else {
-					for (var ind = 1; ind < m.length; ind++) {
-						if (context.groups && (ind-1) < context.groups.length) {
-							context[context.groups[ind-1]] = m[ind]
+
+				let params = route.path.exec(path);
+				let options = (route.options || (() => ({ render: {}, serving: {} })))(params);
+
+				return reduceToPromise(route.context || {}, params, this.db)
+					.then((context) => {
+						logger.info(`Building ${path}`);
+						context.user = user;
+						let content = this.renderPage(route.page, context);
+
+						if (route.cache) {
+							logger.ok(`Caching ${path}`);
+							return fs.writeFile(cachePath, content)
+								.then(() => content);
 						}
-					}
-					if (req.user) context.user = req.user
-					var out = this.renderPath(context)
-					res.send(out)
-					return
-				}
-			}
-		}
-		next()
-	},
-	fourohfour: function(req, res, next) {
-		var loc = path.join(this.__dirname, config.paths.render)
-		//Inelegent, but best I could come up with in my tired state
-		promiseFile(loc, RENDER_ROOT_STR+".404").then(function(file) {
-			res.status(404)
-			res.send(file.data)
-		}).catch(function(err) {
-			res.status(666) // this is not a valid http code.  please resubmit.
-			logger.error("[error]", "Error:", err)
-			res.send("Error Recursion too deep. Please brace for the apocalypse.")
-		}).catch(crash)
+
+						return content;
+					})
+					.then((content) => { // the render succeeded
+						if (res !== undefined) {
+							logger.ok(`Sending ${path}`);
+							res.type("text/html"); // TODO
+							res.send(content);
+						}
+					})
```

Comment:
can we put the file write after the response? That way if something weird happens with fileIO (as it is want to), the server can still send a response